File: Samples/XamlListView/cs/Samples/TapOnTheEdge/EdgeTappedListView.cs
using System;
using Windows.UI.Input;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Input;
using Windows.UI.Xaml.Media;
using Windows.UI.Xaml.Shapes;

namespace ListViewSample
{
    public class ListViewEdgeTappedEventArgs : EventArgs
    {
        private ListViewItem listViewItem;
        public ListViewItem ListViewItem { get { return listViewItem; } }
        public ListViewEdgeTappedEventArgs(ListViewItem listViewItem)
        {
            this.listViewItem = listViewItem;
        }
    }
    public class EdgeTappedListView : ListView
    {
        private const double HIT_TARGET = 32;
        private const double VISUAL_INDICATOR_WIDTH = 12;
        private const string VISUAL_INDICATOR_NAME = "VisualIndicator";
        private ListViewItem listViewItemHighlighted;
        private Rectangle visualIndicator;

        public delegate void ListViewEdgeTappedEventHandler(ListView sender, ListViewEdgeTappedEventArgs e);
        public event ListViewEdgeTappedEventHandler ItemLeftEdgeTapped;
        public Brush LeftEdgeBrush { get; set; }
        public bool IsItemLeftEdgeTapEnabled { get; set; }

        public EdgeTappedListView()
        {
            this.ContainerContentChanging += OnContainerContentChanging;
            listViewItemHighlighted = null;
            LeftEdgeBrush = Application.Current.Resources["SystemControlHighlightAltListAccentLowBrush"] as Brush;
        }
        private void OnContainerContentChanging(ListViewBase sender, ContainerContentChangingEventArgs args)
        {
            if (args.InRecycleQueue)
            {
                ListViewItem lvi = (args.ItemContainer as ListViewItem);
                if (lvi != null)
                {
                    UIElement element = VisualTreeHelper.GetChild(lvi, 0) as UIElement;
                    if (element != null)
                    {
                        element.PointerPressed -= OnPointerPressed;
                        element.PointerReleased -= OnPointerReleased;
                        element.PointerCaptureLost -= OnPointerCaptureLost;
                        element.PointerExited -= OnPointerExited;
                    }
                }
            }
            else if (args.Phase == 0)
            {
                ListViewItem lvi = (args.ItemContainer as ListViewItem);
                if (null != lvi)
                {
                    UIElement element = VisualTreeHelper.GetChild(lvi, 0) as UIElement;
                    if (null != element)
                    {
                        element.PointerPressed += OnPointerPressed;
                        element.PointerReleased += OnPointerReleased;
                        element.PointerCaptureLost += OnPointerCaptureLost;
                        element.PointerExited += OnPointerExited;
                    }
                }
            }
        }
        void OnPointerPressed(object sender, PointerRoutedEventArgs e)
        {
            if (IsItemLeftEdgeTapEnabled)
            {
                // This conditional was commented to enable this on non-Mobile devices.
                // var qualifiers = Windows.ApplicationModel.Resources.Core.ResourceContext.GetForCurrentView().QualifierValues;
                // if (qualifiers.ContainsKey("DeviceFamily") && qualifiers["DeviceFamily"] == "Mobile")
                {
                    if (e.Pointer.PointerDeviceType == Windows.Devices.Input.PointerDeviceType.Touch)
                    {
                        UIElement element = (sender as UIElement);
                        if (null != element)
                        {
                            PointerPoint pointerPoint = e.GetCurrentPoint(element);
                            if (pointerPoint.Position.X <= HIT_TARGET)
                            {
                                listViewItemHighlighted = VisualTreeHelper.GetParent(element) as ListViewItem;
                                ShowVisual();
                            }
                        }
                    }
                }
            }
        }
        void OnPointerReleased(object sender, PointerRoutedEventArgs e)
        {
            if (listViewItemHighlighted != null)
            {
                ClearVisual();
            }
        }
        private void OnPointerCaptureLost(object sender, PointerRoutedEventArgs e)
        {
            ClearVisual();
        }
        void OnPointerExited(object sender, PointerRoutedEventArgs e)
        {
            ClearVisual();
        }
        private void ShowVisual()
        {
            if (listViewItemHighlighted != null)
            {
                visualIndicator = listViewItemHighlighted.FindName("VISUAL_INDICATOR_NAME") as Rectangle;
                if (visualIndicator == null)
                {
                    visualIndicator = new Rectangle();
                    visualIndicator.Name = VISUAL_INDICATOR_NAME;
                    visualIndicator.Height = listViewItemHighlighted.ActualHeight;
                    visualIndicator.HorizontalAlignment = HorizontalAlignment.Left;
                    visualIndicator.Width = VISUAL_INDICATOR_WIDTH;
                    visualIndicator.Fill = LeftEdgeBrush;
                    visualIndicator.Margin = new Thickness(-(listViewItemHighlighted.Padding.Left), 0, 0, 0);
                    Panel panel = listViewItemHighlighted.ContentTemplateRoot as Panel;
                    if (panel != null)
                    {
                        if (panel is Grid)
                        {
                            visualIndicator.SetValue(Grid.RowSpanProperty, (panel as Grid).RowDefinitions.Count);
                        }
                        panel.Children.Add(visualIndicator);
                    }
                }
                else
                {
                    visualIndicator.Opacity = 1;
                }
            }
        }
        private void ClearVisual()
        {
            if (listViewItemHighlighted != null)
            {
                if (visualIndicator != null)
                {
                    visualIndicator.Opacity = 0;
                    if (this.ItemLeftEdgeTapped != null)
                    {
                        ItemLeftEdgeTapped(this, new ListViewEdgeTappedEventArgs(listViewItemHighlighted));
                    }
                }
                listViewItemHighlighted = null;
            }
        }
    }
}

File: Samples/XamlListView/cppwinrt/EdgeTappedListView.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "EdgeTappedListView.h"
#include "EdgeTappedListView.g.cpp"
#include "ListViewEdgeTappedEventArgs.g.cpp"
#include <mutex>

using namespace winrt::Windows::UI::Input;
using namespace winrt::Windows::UI::Xaml;
using namespace winrt::Windows::UI::Xaml::Controls;
using namespace winrt::Windows::UI::Xaml::Input;
using namespace winrt::Windows::UI::Xaml::Media;
using namespace winrt::Windows::UI::Xaml::Shapes;

namespace winrt::SDKTemplate::implementation
{
    Windows::UI::Xaml::DependencyProperty EdgeTappedListView::s_containerEventTokensProperty{ nullptr };

    void EdgeTappedListView::RegisterDependencyPropertiesOnce()
    {
        static std::once_flag flag;
        std::call_once(flag, [] {
            s_containerEventTokensProperty = DependencyProperty::RegisterAttached(
                L"ContainerEventTokens", xaml_typename<IInspectable>(),
                xaml_typename<SDKTemplate::EdgeTappedListView>(), nullptr);
            });
    }

    EdgeTappedListView::EdgeTappedListView()
    {
        RegisterDependencyPropertiesOnce();
        ContainerContentChanging({ get_weak(), &EdgeTappedListView::OnContainerContentChanging });
        leftEdgeBrush = Application::Current().Resources().Lookup(box_value(L"SystemControlHighlightAltListAccentLowBrush")).as<Brush>();
    }

    void EdgeTappedListView::OnContainerContentChanging(ListViewBase const&, ContainerContentChangingEventArgs const& args)
    {
        if (args.InRecycleQueue())
        {
            UIElement element = VisualTreeHelper::GetChild(args.ItemContainer(), 0).try_as<UIElement>();
            if (element != nullptr)
            {
                IInspectable prop = element.ReadLocalValue(s_containerEventTokensProperty);
                if (prop != nullptr)
                {
                    element.ClearValue(s_containerEventTokensProperty);
                    ContainerEventTokens* tokens = get_self<ContainerEventTokens>(prop);
                    element.PointerPressed(tokens->pointerPressedToken);
                    element.PointerReleased(tokens->pointerReleasedToken);
                    element.PointerCaptureLost(tokens->pointerCaptureLostToken);
                    element.PointerExited(tokens->pointerExitedToken);
                }
            }
        }
        else if (args.Phase() == 0)
        {
            UIElement element = VisualTreeHelper::GetChild(args.ItemContainer(), 0).try_as<UIElement>();
            if (element != nullptr)
            {
                com_ptr<ContainerEventTokens> tokens = make_self<ContainerEventTokens>();
                element.SetValue(s_containerEventTokensProperty, *tokens);
                tokens->pointerPressedToken = element.PointerPressed({ get_weak(), &EdgeTappedListView::Element_PointerPressed });
                tokens->pointerReleasedToken = element.PointerReleased({ get_weak(), &EdgeTappedListView::Element_PointerReleased });
                tokens->pointerCaptureLostToken = element.PointerCaptureLost({ get_weak(), &EdgeTappedListView::Element_PointerCaptureLost });
                tokens->pointerExitedToken = element.PointerExited({ get_weak(), &EdgeTappedListView::Element_PointerExited });
            }
        }
    }

    void EdgeTappedListView::Element_PointerPressed(IInspectable const& sender, PointerRoutedEventArgs const& e)
    {
        if (IsItemLeftEdgeTapEnabled())
        {
            if (e.Pointer().PointerDeviceType() == Windows::Devices::Input::PointerDeviceType::Touch)
            {
                UIElement element = sender.as<UIElement>();
                PointerPoint pointerPoint = e.GetCurrentPoint(element);
                if (pointerPoint.Position().X <= HIT_TARGET)
                {
                    listViewItemHighlighted = VisualTreeHelper::GetParent(element).as<ListViewItem>();
                    ShowVisual();
                }
            }
        }
    }

    void EdgeTappedListView::Element_PointerReleased(IInspectable const&, PointerRoutedEventArgs const&)
    {
        if (listViewItemHighlighted != nullptr)
        {
            ClearVisual();
        }
    }

    void EdgeTappedListView::Element_PointerCaptureLost(IInspectable const&, PointerRoutedEventArgs const&)
    {
        ClearVisual();
    }

    void EdgeTappedListView::Element_PointerExited(IInspectable const&, PointerRoutedEventArgs const&)
    {
        ClearVisual();
    }

    void EdgeTappedListView::ShowVisual()
    {
        if (listViewItemHighlighted != nullptr)
        {
            visualIndicator = listViewItemHighlighted.FindName(VISUAL_INDICATOR_NAME).as<Rectangle>();
            if (visualIndicator == nullptr)
            {
                visualIndicator = Rectangle();
                visualIndicator.Name(VISUAL_INDICATOR_NAME);
                visualIndicator.Height(listViewItemHighlighted.ActualHeight());
                visualIndicator.HorizontalAlignment(HorizontalAlignment::Left);
                visualIndicator.Width(VISUAL_INDICATOR_WIDTH);
                visualIndicator.Fill(leftEdgeBrush);
                visualIndicator.Margin(Thickness{ -(listViewItemHighlighted.Padding().Left), 0, 0, 0 });
                Panel panel = listViewItemHighlighted.ContentTemplateRoot().try_as<Panel>();
                if (panel != nullptr)
                {
                    Grid grid = panel.try_as<Grid>();
                    if (grid != nullptr)
                    {
                        visualIndicator.SetValue(Grid::RowSpanProperty(), box_value(grid.RowDefinitions().Size()));
                    }
                    panel.Children().Append(visualIndicator);
                }
            }
            else
            {
                visualIndicator.Opacity(1.0);
            }
        }
    }

    void EdgeTappedListView::ClearVisual()
    {
        if (listViewItemHighlighted != nullptr)
        {
            if (visualIndicator != nullptr)
            {
                visualIndicator.Opacity(0.0);
                if (itemLeftEdgeTappedEvent)
                {
                    itemLeftEdgeTappedEvent(*this, make<ListViewEdgeTappedEventArgs>(listViewItemHighlighted));
                }
            }
            listViewItemHighlighted = nullptr;
        }
    }
}
